### DISCUSSION TITLE
Mapstruct: Ambiguous mapping methods found for mapping property

### DISCUSSION THREAD

**Discussion by mster429** [mapstruct/mapstruct#3023] [Q&A] [answered]
I am facing this [issue](https://stackoverflow.com/q/73767080/7882445) with mapstruct mapping in my spring boot application after I upgrade the sprin boot version. Any help is appreciated. I am stuck with this.

```
spring boot version: 2.7.2
mapstruct version: 1.3.1.Final
mapstruct-processor version: 1.3.1.Final
lombok version: 1.18.24
lombok mapstruct binding version: 0.2.0
```

**Comment by chenzijia12300** [comment, marked as answer]
Hi, I guess you can read this part of the [documentation](https://mapstruct.org/documentation/stable/reference/html/#selection-based-on-qualifiers)

**Reply by mster429** [reply]
Thanks for you reply. I have gone through that part. But I am facing a very specific issue when trying to map using JPARepository.

`Ambiguous mapping methods found for mapping property "(@javax.validation.constraints.NotNull(message="This field is required") :: java.lang.Short) leadGenLoanType" to (@javax.validation.constraints.NotNull :: com.talos.dealt.domain.LeadGenGeneralLoanType): com.talos.dealt.domain.LeadGenGeneralLoanType com.talos.dealt.repository.LeadGenGeneralLoanTypeRepository.getOne(java.lang.Short id), com.talos.dealt.domain.LeadGenGeneralLoanType com.talos.dealt.repository.LeadGenGeneralLoanTypeRepository.getById(java.lang.Short id), com.talos.dealt.domain.LeadGenGeneralLoanType com.talos.dealt.repository.LeadGenGeneralLoanTypeRepository.getReferenceById(java.lang.Short id).`

```
@Repository
public interface LeadGenInvestmentLoanTypeRepository extends JpaRepository<LeadGenInvestmentLoanType, Short>{
}
```

and the implementation for getOne, getById and getReferenceByUd in Simple.JpaRepository

```
    /** @deprecated */
    @Deprecated
    public T getOne(ID id) {
        return this.getReferenceById(id);
    }

    /** @deprecated */
    @Deprecated
    public T getById(ID id) {
        return this.getReferenceById(id);
    }

    public T getReferenceById(ID id) {
        Assert.notNull(id, "The given id must not be null!");
        return this.em.getReference(this.getDomainClass(), id);
    }
```
Is there any way to guide mapstruct to use this `getReferenceById` method to map my TypeA to TypeB ?

**Reply by mster429** [reply]
```
@Mapper(componentModel = "spring",
    uses = {LeadGenInvestmentLoanTypeRepository.class, LeadGenInformationMapper.class,
        DocumentMapper.class, AddressMapper.class},
    unmappedTargetPolicy = ReportingPolicy.ERROR)
public abstract class SmsfLeadGenMapper {

    @Mappings({
        @Mapping(target = "leadGenInvestmentLoanType", source = "request.leadGenLoanType")
    })
    public abstract void sourceToDestination(final SmsfLeadGenRequest request,
                                             final User user,
                                             final @MappingTarget SmsfLeadGenInput smsfLeadGenInput);
```

**Reply by mster429** [reply]
Any help is appreciated. I am stuck with this.

**Reply by chenzijia12300** [reply]
You can use the `@Named` annotation with the `qualifiedByName` attribute of the `@Mapping` annotation to specify a specific mapping method
I think the documentation I mentioned above will solve your problem. You can try reading the official [documentation](https://mapstruct.org/documentation/stable/reference/html/#selection-based-on-qualifiers) carefully

**Comment by ro0sterjam** [comment]
You have three methods all with the same param and return types. MapStruct has no idea which of the three to use when mapping from an `ID` to some generic type `T`. Either mark your methods with `@Named` and use that to specify which method to use in the mapping, or don't use MapStruct in this way. This is personally a bit odd for me to see a repository being used as a mapper.
